### PR TITLE
Function for Grays

### PR DESCRIPTION
**Pull request by HuntJSparra**
Partially responds to #3347:

- Adds grey/gray function similar to magma
- Adds greys256 palette
- Adds warning for when more than 256 colors are requested since more than 256 colors is not always possible

### CODE REVIEW

File: bokeh/palettes.py [hunk 1]
```diff
@@ -432,7 +455,7 @@
     "Greens3",    "Greens4",    "Greens5",    "Greens6",    "Greens7",    "Greens8",    "Greens9",
     "Oranges3",   "Oranges4",   "Oranges5",   "Oranges6",   "Oranges7",   "Oranges8",   "Oranges9",
     "Reds3",      "Reds4",      "Reds5",      "Reds6",      "Reds7",      "Reds8",      "Reds9",
-    "Greys3",     "Greys4",     "Greys5",     "Greys6",     "Greys7",     "Greys8",     "Greys9",
+    "Greys3",     "Greys4",     "Greys5",     "Greys6",     "Greys7",     "Greys8",     "Greys9",                                   "Greys256"
```
Comment (canavandl) [inline bokeh/palettes.py]: Missing a comma. I'd also remove the extra whitespace (not try to align with vertically with other 256 palettes) and just put it in the next sequential column after `Greys9`.

**Comment by canavandl** [comment]
Example code:

```python
import numpy as np
from matplotlib.mlab import bivariate_normal

from bokeh.plotting import figure, output_file, show
from bokeh.models import Range1d, LogColorMapper

output_file('image.html')

N = 100
X, Y = np.mgrid[-3:3:complex(0, N), -2:2:complex(0, N)]

# A low hump with a spike coming out of the top right.  Needs to have
# z/colour axis on a log scale so we see both hump and spike.  linear
# scale only shows the spike.
Z1 = bivariate_normal(X, Y, 0.1, 0.2, 1.0, 1.0) +  \
    0.1 * bivariate_normal(X, Y, 1.0, 1.0, 0.0, 0.0)

image = Z1 / Z1.min() #normalize to increase log range

mapper = LogColorMapper(palette="Greys256")

p = figure(x_range=Range1d(0,1), y_range=Range1d(0,1))

p.image(image=[image],
        color_mapper=mapper,
        # palette='Spectral11',
        dh=[1.0],
        dw=[1.0],
        x=[0],
        y=[0])

show(p)
```

yields:

screen shot 2016-06-29 at 3 14 08 pm

compares well to other colormaps in https://github.com/bokeh/bokeh/pull/4549#issuecomment-226641972

**Comment by canavandl** [comment]
LGTM - I think the greyscale levels (rate it goes from black to white) seem appropriate.

@HuntJSparra - If you can try to include some example code to run or screenshots (especially if there aren't accompanying unit tests), it makes it easier for others to test. Next time :)

**Comment by HuntJSparra** [comment]
Will do in the future. Thanks for looking over everything.

**Comment by bryevdv** [comment]
@HuntJSparra Let 's take this opportunity to add tests for `bokeh.palettes`. There is a test file in `bokeh/tests/test_palettes`  it would be good to ensure that: 
* The expected palettes are present in the module
* The expected cmap functions are present in the module
Possibly going further:
* The palettes present all have the expected lengths
* Tests for the proper function of the generic cmap generator

File: bokeh/palettes.py [hunk 2]
```diff
@@ -488,6 +511,8 @@
 
 def _cmap_func_generator(name, cmap):
     def func(n):
+        if n>256:
```
Comment (bryevdv) [inline bokeh/palettes.py]: `>=256` or `>255` I think?

Comment (bryevdv) [inline bokeh/palettes.py]: Also I would say raise a `ValueError`. But I think this function needs a new name. This 256 restriction only applies to certain specific cases. Something extremely descriptive maybe: `_linear_cmap_from_256_colors`

**Comment by bryevdv** [comment]
@HuntJSparra some small comments

**Comment by bryevdv** [comment]
superceded by #4751

**Pull request closed without merging**